compiler/transf.nim

```diff
@@ -780,6 +780,43 @@ proc commonOptimizations*(g: ModuleGraph; c: PSym, n: PNode): PNode =
     else:
       result = n
 
+proc hoistParamsUsedInDefault(c: PTransf, call, letSection, defExpr: PNode): PNode =
```

Comment:
``semcall`` rewrites the call expression to what is actually meant. I don't know why this hoisting is done in ``transf``. Macros benefit from an AST that is not subtle to resolve too.

Reply:
My reasoning was that it would be more complicated to handle all the different ways calls are processed in sem (dot calls, indirect calls, etc), but I could give it a try. I'm not entirely convinced that making such rewrites visible to the user code is the right thing (e.g. 
 to`proc.getImpl`), but you might have a point.

BTW, this pull request doesn't cover yet the usage of such params in templates and macros (the semantics there will have to be a bit different I think). I considered making it an error for now, but forgot to add the code for this.

Reply:
> I'm not entirely convinced that making such rewrites visible to the user code is the right thing (e.g. toproc.getImpl), but you might have a point.

That's what Nim currently does for simpler default parameters and it's helpful in macros that use ``typed``.